### PR TITLE
Added an adapter for FastAPI

### PR DESCRIPTION
**Pull request by LeOndaz**
This PR introduces an adapter for FastAPI. I'll create an example and PR it later.

### CODE REVIEW

**Comment by jensens** [comment]
Thanks, this is great! Could you provide an example usage with test? like here: https://github.com/authomatic/authomatic/tree/master/examples

**Comment by LeOndaz** [comment]
> Thanks, this is great! Could you provide an example usage with test? like here: https://github.com/authomatic/authomatic/tree/master/examples

Thank you. I'll soon enough

**Comment by LeOndaz** [comment]
@jensens I think I'm having some problems running the tests, after I installed the deps, I run `tox` in the project directory and 
```
py3 run-test: commands[0] | pylint --errors-only --ignore=six.py authomatic
************* Module authomatic.extras.flask
authomatic/extras/flask.py:47:8: E0237: Assigning to attribute 'modified' not defined in class slots (assigning-non-slot)
ERROR: InvocationError for command /home/leondaz/contrib/authomatic/.tox/py3/bin/pylint --errors-only --ignore=six.py authomatic (exited with code 2)
____________________________________________________________________________________________________________ summary ____________________________________________________________________________________________________________
ERROR:   py3: commands failed
```
I'm running Python 3.8

**Comment by LeOndaz** [comment]
@jensens I understand the error, but I've assumed that this error happened only to me because if authomatic is on GitHub with TravisCI then the tests should be passing right?

**Comment by mrichar1** [comment]
I haven't had a deep dig, but it looks like the authomatic code calls `session.modified = True` (where`session` is imported from `flask`. My guess is that the `flask` api has changed in some way. I'll have a read up on the flask docs and see if this line needs modified or removed.

**Comment by mrichar1** [comment]
OK so it looks like this is still the 'right thing to do' according to the Flask docs: https://flask.palletsprojects.com/en/2.0.x/api/#sessions

It looks like this is probably a bug in `flask` (something like they define `__slots__` but don't have `modified` in it). It's probably safe to have a pylint 'ignore' for this line, but we should open a bug with Flask and get them to confirm this.

@LeOndaz if you want to work round this for now, just chang the last line in `authomatic/extras/flask.py` to be:

```
        session.modified = True  # pylint: disable=assigning-non-slot
```

(but don't commit this in any PRs you make, and we'll hopefully resolve it separately).

**Comment by mrichar1** [comment]
( @jensens Raised with Flask in https://github.com/pallets/flask/issues/4219 - will see what they say)

**Comment by LeOndaz** [comment]
@mrichar1 Thank you for your efforts, not gonna lie, I removed flask support when I got the error :joy:

**Comment by LeOndaz** [comment]
@jensens @mrichar1 Kindly, review when you have time. And thanks again for your efforts, I appreciate.

**Comment by LeOndaz** [comment]
That's it for today.  Sorry for missing the formatting thing, I didn't notice that there's an unformatted file.

**Comment by mrichar1** [comment]
Flask have confirmed the pylint issue is an upstream pylint false-positive, so have pushed the `pylint disable` comment to `master` to hide this error.

**Comment by LeOndaz** [comment]
@mrichar1 I've fixed the conflict, you've added the comment and I've disabled pylint for the line

**Comment by LeOndaz** [comment]
@mrichar1 Hello, can you review and merge if it's okay with you?

**Comment by mrichar1** [comment]
We've just been making changes to re-enable automatic testing on PRs and commits - I've just merged these chanegs in master onto this PR to get these tests running. Will then start a code review when they complete.

File: tests/functional_tests/test_providers.py [hunk 1]
```diff
@@ -45,6 +45,21 @@
...
+    def create_command(self):
+        return [
+            self.executable,
+            '{}:app'.format(self.path),
+            '--host {}'.format(self.host),
+            '--port {}'.format(self.port),
+            '--reload',
+        ]
+
+
```
Comment (mrichar1) [inline tests/functional_tests/test_providers.py:62]: This should probably be a separate commit to the liveandletdie project (which is part of the authomatic org, so we can accept the PR easily enough).

File: examples/fastapi/functional_test/app.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import authomatic
+from fastapi import FastAPI
+from fastapi.responses import PlainTextResponse
+from authomatic.adapters import DjangoAdapter
```
Comment (mrichar1) [inline examples/fastapi/functional_test/app.py:4]: Shouldn't this be FastAPIAdapter? (see also line 21)

File: authomatic/adapters.py [hunk 3]
```diff
@@ -280,3 +280,36 @@ def set_header(self, key, value):
 
     def set_status(self, status):
         self.response.status = status
+
```
Comment (mrichar1) [inline authomatic/adapters.py:283]: Please add a Sphinx doc `autoclass` entry for this adapter in the main docstring (aorund line 30)

**Comment by LeOndaz** [comment]
@mrichar1 Perfect, Thanks for your efforts. we've now 4 checks that passed.

File: examples/fastapi/functional_test/app.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+@app.route('/login/{provider_name}', methods=['GET', 'POST'])
+def login(request, provider_name):
+    response = PlainTextResponse()
+    result = authomatic.login(DjangoAdapter(request, response), provider_name)
+
+    if result:
+        return fixtures.render_login_result('fastapi', result)
+
+    return response
+
```
Comment (mrichar1) [inline examples/fastapi/functional_test/app.py:27]: Blank line at end, containing whitespace

File: authomatic/adapters.py [hunk 5]
```diff
@@ -280,3 +280,36 @@ def set_header(self, key, value):
...
+
+class FastAPIAdapter(BaseAdapter):
```
Comment (mrichar1) [inline authomatic/adapters.py:285]: Missing class docstring

**Review by mrichar1 — CHANGES_REQUESTED** [review]

File: tests/functional_tests/test_providers.py (hunk 1 above)
Comment (LeOndaz) [reply tests/functional_tests/test_providers.py:62]: @mrichar1 Yes, it should, however, I saw that it's 2 years old project and no longer maintained and this is why I added it here, however, we can directly import it from liveandletdie and I'll handle the PR there atm, we can skip this one. Let me PR there real quick

File: examples/fastapi/functional_test/app.py (hunk 2 above)
Comment (LeOndaz) [reply examples/fastapi/functional_test/app.py:4]: Yes, mb, it still passes the tests, I solved it and I'm gonna commit after you review liveandletdie PR because the next commit here depends on it

File: examples/fastapi/functional_test/app.py (hunk 4 above)
Comment (LeOndaz) [reply examples/fastapi/functional_test/app.py:27]: Resolved

File: authomatic/adapters.py (hunk 5 above)
Comment (LeOndaz) [reply authomatic/adapters.py:285]: Resolved

File: tests/functional_tests/test_providers.py (hunk 1 above)
Comment (mrichar1) [reply tests/functional_tests/test_providers.py:62]: That's a new `liveandletdie` released.

Comment (LeOndaz) [reply tests/functional_tests/test_providers.py]: @mrichar1 Great, now I pushed here and fixed the issues. Have a look.

**Review by mrichar1 — APPROVED** [review]
LGTM :+1:

**Comment by mrichar1** [comment]
I had missed that while fastAPI was added to the test stack, it wasn't actually being selected in the test config. I've updated this setting in your branch, which should cause it to run - though the tests that require auth secrets will be skipped. I'll run these offline and feed back if there are any issues.

**Comment by mrichar1** [comment]
So the test with secrets failed as `liveandletdie` is failing to start `FastAPIServer` properly. It looks like this is down to `liveandletdie` not starting `uvicorn`/the example app properly.  This is down to a few things:

* `uvicorn` needs a `app-dir` parameter to define the path that the module is loaded from - so the long path needs splitting out into pth and module.
* There's no `app` or equivalent 'start-point' in the example app for `uvicorn to load (i.e. `main:main` or equivalent).
* There *might* be issues if these methods aren't defined as asynchronous.

Can you look at getting the test suite updated in `liveandletdie` to start uvicorn and run the test app correctly? This will then let us fully test this PR.

**Comment by LeOndaz** [comment]
@mrichar1 
I'll be investigating and get back to you

**Comment by LeOndaz** [comment]
@mrichar1 hopefully you're still alive 😅

File: examples/fastapi/app/app.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+                        'https://www.googleapis.com/oauth2/v1/userinfo?alt=json'
+                    )
+
+                    if provider_response.status == 200:
+                        response.body += 'Hello {}'.format(provider_response.data.name).encode('utf-8')
+
+                if provider.name == 'github':
+                    response.body += 'You\'re now logged in with GitHub.<br />'.encode('utf-8')
+
+                    provider_response = provider.access(user.credentials, 'https://api.github.com/user', {
```
Comment (jensens) [inline examples/fastapi/app/app.py:53]:
```suggestion
                    provider_response = provider.access('https://api.github.com/user', headers={
```

Comment (jensens) [reply examples/fastapi/app/app.py:53]: here the signature and data passed are wrong.

File: examples/fastapi/app/app.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64)'
+                                      ' AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.131 Safari/537.36'
+                    })
+
+                    if provider_response.status == 200:
+                        response.body += 'Hello {}'.format(provider_response.data.name).encode('utf-8')
```
Comment (jensens) [inline examples/fastapi/app/app.py:59]: dict access here, not dot access
```suggestion
                        response.body += 'Hello {}'.format(provider_response.data['name']).encode('utf-8')

```

File: examples/fastapi/app/app.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+                if provider.name == 'google':
+                    response.body += 'You\'re now logged in with Google.<br />'.encode('utf-8')
+
+                    provider_response = provider.access(
+                        user.credentials,
+                        'https://www.googleapis.com/oauth2/v1/userinfo?alt=json'
+                    )
+
+                    if provider_response.status == 200:
+                        response.body += 'Hello {}'.format(provider_response.data.name).encode('utf-8')
```
Comment (jensens) [inline examples/fastapi/app/app.py:48]: dict access here, not dot access
```suggestion
                        response.body += 'Hello {}'.format(provider_response['data'].name).encode('utf-8')

```

**Review by jensens — CHANGES_REQUESTED** [review]
findings

**Comment by LeOndaz** [comment]
Feel free to commit in this PR and merge after you are done, I'm no longer interested in authomatic as it's unmaintained so I won't be exerting effort for it

**Comment by jensens** [comment]
@mrichar1 this is green now. So, nothing stops us from merging, right?

**Comment by mrichar1** [comment]
Yes - I finally managed to get tests and releases working again (!), so if it's passing then it should be fine to merge.

To make a new release, update the version in `setup.cfg`, and push to `master` tagged with the version number, and a new release should appear on PyPi. :crossed_fingers:

**Comment by jensens** [comment]
> To make a new release, update the version in `setup.cfg`, and push to `master` tagged with the version number, and a new release should appear on PyPi. crossed_fingers
> 
> (A push to `master` will run the full set of tests, including the functional oauth2 tests against Facebook, so might be best to push, then if that passes, push the tag after).

The new release process has a glitch, see #225

**Comment by mrichar1** [comment]
It looks like the 'full' run of this the test suite for this PR is failing, with none of the FastAPI tests passing. The issue seems to resolve around `liveandletdie` not starting the FastAPI server correctly:

`liveandletdie.LiveAndLetDieError: FastAPIServer server http://authomatic.org:443/ didn't start in specified timeout 10.0 seconds!`

I can't remember what state `FastAPI` support in `liveandletdie` got to, if at all?

If we think that this code is valid, but the test is failing due to a lack of support in `liveandletdie` then we can leave this commit in, but remove `fastapi` from `test_providers.py` to prevent it trying and failing to test it - at least until we can patch `liveandletdie` appropriately...

**Comment by jensens** [comment]
I never dig into `liveandletdie` much. But that's probably the issue.
